src/ast/transform/SemanticChecker.cpp

```diff
@@ -504,51 +504,41 @@ void SemanticCheckerImpl::checkClause(const Clause& clause) {
 }
 
 void SemanticCheckerImpl::checkComplexRule(const std::set<const Clause*>& multiRule) {
-    // each variable appearing in a clause body must also appear:
-    // - at least a second time in the same clause body,
-    // - or at least once in one of the heads of the clause.
-    std::set<std::string> var_in_head;
-    std::map<std::string, const ast::Variable*> report_var_pos;
+    std::map<std::string, int> var_count;
+    std::map<std::string, const ast::Variable*> var_pos;
 
-    for (auto clause : multiRule) {
-        visit(*(clause->getHead()), [&](const ast::Variable& var) { var_in_head.emplace(var.getName()); });
+    auto count_var = [&](const ast::Variable& var) {
+        const auto& varName = var.getName();
+        if (0 == var_count[varName]++) {
+            var_pos[varName] = &var;
+        } else {
+            const auto& PrevLoc = var_pos[varName]->getSrcLoc();
+            const auto& Loc = var.getSrcLoc();
+            if (PrevLoc < Loc) {
+                var_pos[varName] = &var;
+            }
+        }
+    };
+
+    // Count the variable occurrence for the body of a
+    // complex rule only once.
+    // TODO (b-scholz): for negation / disjunction this is not quite
+    // right; we would need more semantic information here.
+    for (auto literal : (*multiRule.begin())->getBodyLiterals()) {
```

Comment:
The first clause returned by `begin()` is not deterministic when compiled with Visual Studio.

Reply:
Oh dear - perhaps it is related to pointers rather than semantic info for ordering in STL containers. We had such issues in the past.

Reply:
Yes. My successive changes in this functions are mostly useless and could be entirely roll-back.

If we really want `begin()` to always return the "first" clause (lexically), then we should provide a dedicated `Comparator` to the `multiRule` set. And that comparator would somehow dig into the pointed clauses, explore the body source-locations and compute the "less" relation.

But my feeling is that there is a more general problem and just relying on the "first" clause is not sufficient to correctly check the whole rule.